File: sdk/ruby/lib/anyvali/interchange/importer.rb
# frozen_string_literal: true

require "json"

module AnyVali
  module Interchange
    module Importer
      module_function

      def import(doc)
        doc = JSON.parse(doc) if doc.is_a?(String)

        # Build definitions first (needed for refs)
        definitions = {}
        if doc["definitions"]
          doc["definitions"].each do |name, node|
            definitions[name] = node_to_schema(node)
          end
        end

        root_schema = node_to_schema(doc["root"])
        context = ValidationContext.new(definitions: definitions)
        [root_schema, context, definitions]
      end

      def import_schema(doc)
        schema, context, definitions = import(doc)
        schema
      end

      def node_to_schema(node)
        kind = node["kind"]

        case kind
        when "string"
          build_string(node)
        when "number", "float32", "float64"
          build_number(node, kind)
        when "int", "int8", "int16", "int32", "int64",
             "uint8", "uint16", "uint32", "uint64"
          build_int(node, kind)
        when "bool"
          build_bool(node)
        when "null"
          NullSchema.new
        when "any"
          AnySchema.new
        when "unknown"
          UnknownSchema.new
        when "never"
          NeverSchema.new
        when "literal"
          LiteralSchema.new(value: node["value"])
        when "enum"
          EnumSchema.new(values: node["values"])
        when "array"
          build_array(node)
        when "tuple"
          build_tuple(node)
        when "object"
          build_object(node)
        when "record"
          build_record(node)
        when "union"
          build_union(node)
        when "intersection"
          build_intersection(node)
        when "optional"
          OptionalSchema.new(schema: node_to_schema(node["schema"]))
        when "nullable"
          build_nullable(node)
        when "ref"
          RefSchema.new(ref: node["ref"])
        else
          raise ValidationError, [
            ValidationIssue.new(
              code: IssueCodes::UNSUPPORTED_SCHEMA_KIND,
              expected: "known schema kind",
              received: kind.to_s
            )
          ]
        end
      end

      def build_string(node)
        constraints = {}
        %w[minLength maxLength pattern startsWith endsWith includes format].each do |k|
          constraints[k] = node[k] if node.key?(k)
        end
        coerce = node["coerce"]
        default_val = node["default"]
        has_default = node.key?("default")
        StringSchema.new(
          constraints: constraints,
          coerce_config: coerce,
          default_value: default_val,
          has_default: has_default
        )
      end

      def build_number(node, kind)
        constraints = {}
        %w[min max exclusiveMin exclusiveMax multipleOf].each do |k|
          constraints[k] = node[k] if node.key?(k)
        end
        coerce = node["coerce"]
        default_val = node["default"]
        has_default = node.key?("default")
        NumberSchema.new(
          kind: kind,
          constraints: constraints,
          coerce_config: coerce,
          default_value: default_val,
          has_default: has_default
        )
      end

      def build_int(node, kind)
        constraints = {}
        %w[min max exclusiveMin exclusiveMax multipleOf].each do |k|
          constraints[k] = node[k] if node.key?(k)
        end
        coerce = node["coerce"]
        default_val = node["default"]
        has_default = node.key?("default")
        IntSchema.new(
          kind: kind,
          constraints: constraints,
          coerce_config: coerce,
          default_value: default_val,
          has_default: has_default
        )
      end

      def build_bool(node)
        coerce = node["coerce"]
        default_val = node["default"]
        has_default = node.key?("default")
        BoolSchema.new(
          coerce_config: coerce,
          default_value: default_val,
          has_default: has_default
        )
      end

      def build_array(node)
        items = node_to_schema(node["items"])
        constraints = {}
        %w[minItems maxItems].each do |k|
          constraints[k] = node[k] if node.key?(k)
        end
        ArraySchema.new(items: items, constraints: constraints)
      end

      def build_tuple(node)
        elements = node["elements"].map { |e| node_to_schema(e) }
        TupleSchema.new(elements: elements)
      end

      def build_object(node)
        props = {}
        (node["properties"] || {}).each do |k, v|
          props[k] = node_to_schema(v)
        end
        required = node["required"] || []
        unknown_keys = node["unknownKeys"] || "reject"
        ObjectSchema.new(
          properties: props,
          required: required,
          unknown_keys: unknown_keys
        )
      end

      def build_record(node)
        values = node_to_schema(node["values"])
        RecordSchema.new(values: values)
      end

      def build_union(node)
        variants = node["variants"].map { |v| node_to_schema(v) }
        UnionSchema.new(variants: variants)
      end

      def build_intersection(node)
        all_of = node["allOf"].map { |v| node_to_schema(v) }
        IntersectionSchema.new(all_of: all_of)
      end

      def build_nullable(node)
        inner = node_to_schema(node["schema"])
        default_val = node["default"]
        has_default = node.key?("default")
        NullableSchema.new(
          schema: inner,
          default_value: default_val,
          has_default: has_default
        )
      end
    end
  end
end

File: sdk/python/src/anyvali/interchange/importer.py
"""Import schemas from AnyVali interchange format."""

from __future__ import annotations

import json
from typing import Any

from ..schemas.any import AnySchema
from ..schemas.array import ArraySchema
from ..schemas.base import BaseSchema, CoercionConfig
from ..schemas.bool import BoolSchema
from ..schemas.enum import EnumSchema
from ..schemas.integer import (
    Int8Schema,
    Int16Schema,
    Int32Schema,
    Int64Schema,
    IntSchema,
    Uint8Schema,
    Uint16Schema,
    Uint32Schema,
    Uint64Schema,
)
from ..schemas.intersection import IntersectionSchema
from ..schemas.literal import LiteralSchema
from ..schemas.never import NeverSchema
from ..schemas.null import NullSchema
from ..schemas.nullable import NullableSchema
from ..schemas.number import Float32Schema, Float64Schema, NumberSchema
from ..schemas.object import ObjectSchema
from ..schemas.optional import OptionalSchema
from ..schemas.record import RecordSchema
from ..schemas.ref import RefSchema
from ..schemas.string import StringSchema
from ..schemas.tuple import TupleSchema
from ..schemas.union import UnionSchema
from ..schemas.unknown import UnknownSchema
from ..types import AnyValiDocument


def import_schema(source: dict[str, Any] | str) -> BaseSchema[Any]:
    """Import a schema from an AnyVali document dict or JSON string.

    Returns the root schema.
    """
    if isinstance(source, str):
        source = json.loads(source)
    assert isinstance(source, dict)

    doc = AnyValiDocument.from_dict(source)

    # Two-pass definition building to handle recursive refs:
    # Pass 1: Create placeholder RefSchema entries so recursive refs can find them
    definitions: dict[str, BaseSchema] = {}

    # Pass 2: Build actual definitions (refs inside will get lazy resolution)
    for name, node in doc.definitions.items():
        definitions[name] = _import_node(node, definitions)

    root = _import_node(doc.root, definitions)

    # Resolve all ref schemas (including those nested in definitions)
    for defn_schema in definitions.values():
        _resolve_refs(defn_schema, definitions)
    _resolve_refs(root, definitions)

    return root


def _parse_coercion(raw: Any) -> CoercionConfig | None:
    """Parse coercion from interchange format.

    Corpus format: either a single string like "string->int" or "trim",
    or a list like ["trim", "lower"].
    """
    if raw is None:
        return None

    flags: dict[str, bool] = {}

    if isinstance(raw, str):
        tokens = [raw]
    elif isinstance(raw, list):
        tokens = raw
    elif isinstance(raw, dict):
        # Dict format (our own export format)
        return CoercionConfig(
            to_int=raw.get("toInt", False),
            to_number=raw.get("toNumber", False),
            to_bool=raw.get("toBool", False),
            trim=raw.get("trim", False),
            lower=raw.get("lower", False),
            upper=raw.get("upper", False),
        )
    else:
        return None

    for token in tokens:
        if token == "string->int":
            flags["to_int"] = True
        elif token == "string->number":
            flags["to_number"] = True
        elif token == "string->bool":
            flags["to_bool"] = True
        elif token == "trim":
            flags["trim"] = True
        elif token == "lower":
            flags["lower"] = True
        elif token == "upper":
            flags["upper"] = True

    if not flags:
        return None

    return CoercionConfig(
        to_int=flags.get("to_int", False),
        to_number=flags.get("to_number", False),
        to_bool=flags.get("to_bool", False),
        trim=flags.get("trim", False),
        lower=flags.get("lower", False),
        upper=flags.get("upper", False),
    )


def _import_node(node: dict[str, Any], definitions: dict[str, BaseSchema]) -> BaseSchema:
    """Import a single schema node from its dict representation."""
    kind = node.get("kind", "")

    schema = _build_schema(kind, node, definitions)

    # Apply common fields
    if "default" in node:
        schema._has_default = True
        schema._default_value = node["default"]

    if "coerce" in node:
        schema._coercion = _parse_coercion(node["coerce"])

    return schema


def _build_int_kwargs(node: dict[str, Any]) -> dict[str, Any]:
    """Extract numeric constraint kwargs from a node."""
    kw: dict[str, Any] = {}
    if "min" in node:
        kw["min"] = node["min"]
    if "max" in node:
        kw["max"] = node["max"]
    if "exclusiveMin" in node:
        kw["exclusive_min"] = node["exclusiveMin"]
    if "exclusiveMax" in node:
        kw["exclusive_max"] = node["exclusiveMax"]
    if "multipleOf" in node:
        kw["multiple_of"] = node["multipleOf"]
    return kw


def _build_schema(
    kind: str, node: dict[str, Any], definitions: dict[str, BaseSchema]
) -> BaseSchema:
    """Build a schema from a kind string and node dict."""
    if kind == "string":
        return StringSchema(
            min_length=node.get("minLength"),
            max_length=node.get("maxLength"),
            pattern=node.get("pattern"),
            starts_with=node.get("startsWith"),
            ends_with=node.get("endsWith"),
            includes=node.get("includes"),
            format_=node.get("format"),
        )
    elif kind in ("number", "float64"):
        cls = NumberSchema if kind == "number" else Float64Schema
        return cls(**_build_int_kwargs(node))
    elif kind == "float32":
        return Float32Schema(**_build_int_kwargs(node))
    elif kind in ("int", "int64"):
        cls = IntSchema if kind == "int" else Int64Schema
        return cls(**_build_int_kwargs(node))
    elif kind == "int8":
        return Int8Schema(**_build_int_kwargs(node))
    elif kind == "int16":
        return Int16Schema(**_build_int_kwargs(node))
    elif kind == "int32":
        return Int32Schema(**_build_int_kwargs(node))
    elif kind == "uint8":
        return Uint8Schema(**_build_int_kwargs(node))
    elif kind == "uint16":
        return Uint16Schema(**_build_int_kwargs(node))
    elif kind == "uint32":
        return Uint32Schema(**_build_int_kwargs(node))
    elif kind == "uint64":
        return Uint64Schema(**_build_int_kwargs(node))
    elif kind == "bool":
        return BoolSchema()
    elif kind == "null":
        return NullSchema()
    elif kind == "any":
        return AnySchema()
    elif kind == "unknown":
        return UnknownSchema()
    elif kind == "never":
        return NeverSchema()
    elif kind == "literal":
        return LiteralSchema(node["value"])
    elif kind == "enum":
        return EnumSchema(node["values"])
    elif kind == "array":
        items_schema = _import_node(node["items"], definitions)
        return ArraySchema(
            items_schema,
            min_items=node.get("minItems"),
            max_items=node.get("maxItems"),
        )
    elif kind == "tuple":
        # Support both "items" and "elements" keys
        items_raw = node.get("elements") or node.get("items", [])
        items = [_import_node(item, definitions) for item in items_raw]
        return TupleSchema(items)
    elif kind == "object":
        props: dict[str, BaseSchema] = {}
        for pname, pnode in node.get("properties", {}).items():
            props[pname] = _import_node(pnode, definitions)
        required = node.get("required")
        unknown_keys = node.get("unknownKeys", "reject")
        return ObjectSchema(props, required=required, unknown_keys=unknown_keys)
    elif kind == "record":
        value_schema = _import_node(node["values"], definitions)
        return RecordSchema(value_schema)
    elif kind == "union":
        # Support both "schemas" and "variants" keys
        schemas_raw = node.get("variants") or node.get("schemas", [])
        schemas = [_import_node(s, definitions) for s in schemas_raw]
        return UnionSchema(schemas)
    elif kind == "intersection":
        # Support both "schemas" and "allOf" keys
        schemas_raw = node.get("allOf") or node.get("schemas", [])
        schemas = [_import_node(s, definitions) for s in schemas_raw]
        return IntersectionSchema(schemas)
    elif kind == "optional":
        inner = _import_node(node["schema"], definitions)
        return OptionalSchema(inner)
    elif kind == "nullable":
        inner = _import_node(node["schema"], definitions)
        return NullableSchema(inner)
    elif kind == "ref":
        return RefSchema(node["ref"])
    else:
        raise ValueError(f"Unsupported schema kind: {kind!r}")


def _resolve_refs(schema: BaseSchema, definitions: dict[str, BaseSchema]) -> None:
    """Resolve RefSchema instances to their target schemas."""
    if isinstance(schema, RefSchema):
        # Give ref access to the full definitions dict for lazy resolution
        schema.set_definitions(definitions)
        ref_name = schema._ref
        if ref_name.startswith("#/definitions/"):
            ref_name = ref_name[len("#/definitions/"):]
        if ref_name in definitions:
            # Don't eagerly resolve recursive refs -- lazy resolution handles it
            schema.resolve(definitions[ref_name])
    elif isinstance(schema, (OptionalSchema, NullableSchema)):
        _resolve_refs(schema._inner, definitions)
    elif isinstance(schema, ArraySchema):
        _resolve_refs(schema._items, definitions)
    elif isinstance(schema, TupleSchema):
        for item in schema._items:
            _resolve_refs(item, definitions)
    elif isinstance(schema, ObjectSchema):
        for prop_schema in schema._properties.values():
            _resolve_refs(prop_schema, definitions)
    elif isinstance(schema, RecordSchema):
        _resolve_refs(schema._value_schema, definitions)
    elif isinstance(schema, (UnionSchema, IntersectionSchema)):
        for s in schema._schemas:
            _resolve_refs(s, definitions)
